src/utils/address.ts

```diff
@@ -0,0 +1,22 @@
+import type express from 'express'
+import { Env } from '../env.js'
+import Identity from '../lib/services/identity.js'
+import { getAuthorization } from '../lib/utils/shared.js'
+
+export const determineAddress = async (
```

Comment:
can I suggest we memoize the valud of `self` within this file rather than have separate caches elsewhere? That will remove `self` as a parameter and mean we're caching it more effectively.

I'd also rename this to `determineSelfAddress` to be more descriptive about what this is